### PR TITLE
EXOGTN-2145 :Timezone problem when publishing a document from the sta…

### PR DESCRIPTION
**Pull request by mbenrhouma**
…ged state

### CODE REVIEW

File: web/eXoResources/src/main/webapp/javascript/eXo/webui/UICalendar.js [hunk 1]
```diff
@@ -439,6 +440,16 @@
           dateString = dateString.replace("mm", minute);
           dateString = dateString.replace("ss", second);
         }
+        if(this.isDisplayTimeZone)
+        {
+          var timezone=this.currentDate.getTimezoneOffset()/-60;
+          if(timezone>=0){
+            dateString=dateString+"             +0"+Math.abs(timezone)+"00";
```
Comment (thomasdelhomenie) [inline web/eXoResources/src/main/webapp/javascript/eXo/webui/UICalendar.js]: I did not understand the purpose of this spaces... until I tested the PRs :panda_face: 
If I understand well, the goal is to get the browser timezone and submit it when the form is submitted, but without make it visible for the user (the spaces make it hidden in the right of the input text). Am I right ?
If so, it is of course not acceptable.
If we need the timezone, we should for example pass it in a hidden field.

Comment (mbenrhouma) [inline web/eXoResources/src/main/webapp/javascript/eXo/webui/UICalendar.js]: The date format  should be :
 **dd/mm/yyyy**  **hh:mm**  **+0X:00** (at least 1 space should be present between time and time zone).
But until i try to change the time, the format becomes  **dd/mm/yyyy  hh:mm+0X:00** (there is no space between time and time zone).

I found that the datePattern after this   :  
  
```
if (getDatePattern_().contains("a")) {
            setDatePattern_(getDatePattern_().replaceAll("a", ""));
        }
```
become : 

 **"dd/mm/yyyy**  **hh:mm "**  

So, i will try to change this pattern to **"dd/mm/yyyy**  **hh:mm"** to resolve this problem and i will delete spaces from this file

Comment (thomasdelhomenie) [inline web/eXoResources/src/main/webapp/javascript/eXo/webui/UICalendar.js]: Actually, what I meant is that the end user should not see the timezone IMO (the part +0X:00). It is a technical detail. So this information should be sent in another way, not directly in the input field (probably in a hidden field). Or we could also convert the time in UTC in javascript so we will always deal with UTC on server-side.

File: web/eXoResources/src/main/webapp/javascript/eXo/webui/UICalendar.js [hunk 2]
```diff
@@ -36,8 +36,9 @@
 
     hideCalendarHandler : null,
 
-    init : function(field, isDisplayTime, datePattern, value, monthNames) {
+    init : function(field, isDisplayTime, datePattern, value, monthNames,isDisplayTimeZone) {
```
Comment (thomasdelhomenie) [inline web/eXoResources/src/main/webapp/javascript/eXo/webui/UICalendar.js]: I think there is no need to add this boolean. The timezone should be managed in all cases.

File: webui/core/src/main/java/org/exoplatform/webui/form/UIFormDateTimeInput.java [hunk 3]
```diff
@@ -48,7 +47,7 @@
      * Whether to display the full time (with hours, minutes and seconds), not only the date
      */
     private boolean isDisplayTime_;
-
+    private boolean isDisplayTimeZone_=false;
```
Comment (thomasdelhomenie) [inline webui/core/src/main/java/org/exoplatform/webui/form/UIFormDateTimeInput.java]: Is this property still useful ?

File: webui/core/src/main/java/org/exoplatform/webui/form/UIFormDateTimeInput.java [hunk 4]
```diff
@@ -146,7 +158,7 @@ private void formatPattern(Locale locale) {
             setDatePattern_(getDatePattern_().replaceAll("H", "HH"));
         }
         if (getDatePattern_().contains("a")) {
-            setDatePattern_(getDatePattern_().replaceAll("a", ""));
+            setDatePattern_(getDatePattern_().replaceAll(" a", ""));
```
Comment (thomasdelhomenie) [inline webui/core/src/main/java/org/exoplatform/webui/form/UIFormDateTimeInput.java]: Is this change intended ? The condition in the if and the operation are not consistent anymore.

File: web/eXoResources/src/main/webapp/javascript/eXo/webui/UIForm.js [hunk 5]
```diff
@@ -28,12 +28,41 @@
...
+		  var dateElements = [];
+		  var originalValues = [];
+		  try {
+			  elements.each(function() {
+				  if ($(this).val()) {
+					  dateElements.push(this);
+					  originalValues.push($(this).val());
+					  // trasformDateValue will add the timezone of browser into value
+
+					  //var date = new Date($(this).val());
```
Comment (thomasdelhomenie) [inline web/eXoResources/src/main/webapp/javascript/eXo/webui/UIForm.js]: To be removed.

File: webui/core/src/main/java/org/exoplatform/webui/form/UIFormDateTimeInput.java [hunk 6]
```diff
@@ -198,13 +214,17 @@ public void processRender(WebuiRequestContext context) throws Exception {
...
+        w.write(String.valueOf(isDisplayTimeZone_));
+        w.write("\"");
         w.write(");' onkeyup='eXo.webui.UICalendar.show();' name='");
         w.write(getName());
         w.write('\'');
         w.write(" value=\"");
         w.write(value);
         w.write('\"');
         w.write(" onclick='event.cancelBubble = true' onkeydown='eXo.webui.UICalendar.onTabOut(event)'");
+        w.write("date-format=\"dd/MM/yyyy hh:mm:ss\"");
```
Comment (thomasdelhomenie) [inline webui/core/src/main/java/org/exoplatform/webui/form/UIFormDateTimeInput.java]: I don't get where you use this value (except to select all the input fields of date type). Can you explain what you did please ?

Comment (mbenrhouma) [inline webui/core/src/main/java/org/exoplatform/webui/form/UIFormDateTimeInput.java]: Of course, this line mean that this input have a date format and it must be converted to UTC

File: web/eXoResources/src/main/webapp/javascript/eXo/webui/UIForm.js (hunk 5 above)
Comment (mbenrhouma) [inline web/eXoResources/src/main/webapp/javascript/eXo/webui/UIForm.js]: OK

File: web/eXoResources/src/main/webapp/javascript/eXo/webui/UIForm.js [hunk 7]
```diff
@@ -28,12 +28,38 @@
...
+					  originalValues.push($(this).val());
+					  var dateString = $(this).val();
+					  var timezone=new Date().getTimezoneOffset()/-60;
+					  if(timezone>=0){
+						  dateString=dateString+" +0"+Math.abs(timezone)+"00";
+					  }else
+					  {
+						  dateString=dateString+" -0"+Math.abs(timezone)+"00";
+					  }
+					  $(this).val(dateString);
```
Comment (thomasdelhomenie) [inline web/eXoResources/src/main/webapp/javascript/eXo/webui/UIForm.js]: This line will update the input field value, so it will be visible by the end user. It should not.

File: web/eXoResources/src/main/webapp/javascript/eXo/webui/UIForm.js [hunk 8]
```diff
@@ -28,12 +28,38 @@
...
+			  elements.each(function() {
+				  if ($(this).val()) {
+					  dateElements.push(this);
+					  originalValues.push($(this).val());
+					  var dateString = $(this).val();
+					  var timezone=new Date().getTimezoneOffset()/-60;
+					  if(timezone>=0){
+						  dateString=dateString+" +0"+Math.abs(timezone)+"00";
+					  }else
+					  {
```
Comment (mbenrhouma) [inline web/eXoResources/src/main/webapp/javascript/eXo/webui/UIForm.js]: OK

File: web/eXoResources/src/main/webapp/javascript/eXo/webui/UIForm.js [hunk 9]
```diff
@@ -28,12 +28,38 @@
...
-	    ajaxPost(url, queryString, callback);
+	  ajaxPost: function (formElement, callback) {
+		  if (!callback)
+			  callback = null;
+
+		  var elements = $(formElement).find('.dateInput');
+		  var dateElements = [];
+		  var originalValues = [];
+		  try {
+			  elements.each(function() {
```
Comment (mbenrhouma) [inline web/eXoResources/src/main/webapp/javascript/eXo/webui/UIForm.js]: OK

File: web/eXoResources/src/main/webapp/javascript/eXo/webui/UIForm.js (hunk 7 above)
Comment (mbenrhouma) [inline web/eXoResources/src/main/webapp/javascript/eXo/webui/UIForm.js]: finally {
 		  dateElements.forEach(function(dateElement) {
 		  dateElement.value = originalValues.shift();
 	  });
This lines will update the input to the originals value, so it will be not visible by the end user.

Comment (thomasdelhomenie) [inline web/eXoResources/src/main/webapp/javascript/eXo/webui/UIForm.js]: So it means we use the input field as placeholder for variables manipulations. That's not a good practice, it is not efficient and can be visible by end user in slow environments. Data manipulation must not use the DOM elements.

File: web/eXoResources/src/main/webapp/javascript/eXo/webui/UIForm.js [hunk 10]
```diff
@@ -31,10 +31,38 @@
...
 	    var url = formElement.action + "&ajaxRequest=true";
 	    ajaxPost(url, queryString, callback);
 	  },
+
+       getQueryString : function(formElement) {
+	     var elements = $(formElement).find('.dateInput');
+         //if the form contains input of type date input
+		 if (elements.size() > 0) {
+		   //clone the form to avoid modifying real DOM elements
+	       var formElementTimeZone = $(formElement).clone();
```
Comment (thomasdelhomenie) [inline web/eXoResources/src/main/webapp/javascript/eXo/webui/UIForm.js]: Instead of cloning the DOM, you can consider using serializeArray and update the date values. Have a look at : http://stackoverflow.com/a/5075798

File: web/eXoResources/src/main/webapp/javascript/eXo/webui/UIForm.js [hunk 11]
```diff
@@ -31,10 +31,33 @@
...
 	    if (!callback)
 	      callback = null;
-	    var queryString = $(formElement).serialize();
+            var queryString = this.getQueryString(formElement);
 	    var url = formElement.action + "&ajaxRequest=true";
 	    ajaxPost(url, queryString, callback);
 	  },
+
+        getQueryString : function(formElement) {
+        var form = $(formElement).find('.dateInput');
```
Comment (thomasdelhomenie) [inline web/eXoResources/src/main/webapp/javascript/eXo/webui/UIForm.js]: Please indent correctly your code.

**Comment by mbenrhouma** [comment]
Merged in develop
